Updates to units section (#158)
yt-project/yt-4.0-paper: content/40.units_and_quantities.md

## Unitful Arrays and Quantities {#sec:units}

At a basic level, `yt` is an engine for converting data created by a simulation code into a
physically meaningful result. Attaching physical units to simulation data makes it possible to
perform dimensional analysis on the simulation data, adding additional opportunities for catching
errors in a data processing pipeline. In addition, it becomes straightforward to convert data from
one unit system to another.

> In `yt` 4.0 we handle units in a an automatic fashion, with a symbolic units system. Originally part

Author: Done, thanks

of `yt` itself since version 3.0, the unit system has now been split off into a standalone package,

> `unyt`. `unyt` leverages the symbolic math library `sympy` in combination with NumPy. In what

Author: Done

follows, we will refer to `unyt` when describing the unit implementation, unless we are describing a
`yt`-specific extension.

Instead of returning a NumPy `ndarray` when users query `yt` data objects for fields, we return a

> `unyt_array`, a subclass of `ndarray`. `unyt_array` preserves `ndarray`'s array operations,

Author: Done

`ndarray`, `unyt_array` attaches unit metadata to the array data, enabling runtime checking of unit
consistency in arithmetic operations between `unyt_array` instances, and making it trivial to
compose new units using algebraic operations.

As a trivial example, when one queries a data object (here given the generic name `dd`) for the
density field, we get back a `unyt_array`, including both the simulation data for the density field,
and the units of the density field, in this case $\rm{g}/\rm{cm}^3$:

```python
>>> dd['density'] 
unyt_array([4.92e-31, 4.94e-31, 4.93e-31, ...,
            1.12e-25, 1.59e-25, 1.09e-24], 'g/cm**3')
```

One of the nicest aspects of the unit system is that the symbolic algebra for unitful operations is performed automatically by `sympy`:

```python
>>> print(dd['mass']/dd['cell_volume'])
  [4.92e-31 4.94e-31 4.93e-31 ... 
   1.12e-25 1.59e-25 1.09e-24] g/cm**3
```

`unyt_array` is primarily useful for attaching units to NumPy `ndarray` instances. For scalar data,
we have created the `unyt_quantity` class. `unyt_quantity` is a subclass of `unyt_array` with the

> requirement that the "array data" associated with the instance be limited to one element.

Reviewer: I'm having trouble coming up with a better phrasing but this is itching me: `unyt_quantity` is in fact slightly more constrained than this, because it can't have `ndim>0` (but it's possible to construct a one-element array in arbitrary dimensions)

Author: Tried to rephrase it a bit to be more precise

    "pc":   (m_per_pc, dimensions.length, 0.0, r"\rm{pc}", True), 
    "degC": (1.0, dimensions.temperature, -273.15, r"^\circ\rm{C}", True),
}
```

Note that we only define **atomic** units here. There should be no operations or SI prefixes in the
registry symbol strings. When we parse non-atomic units like `Msun/Mpc**3`, we use the registry to
look up the symbols. The unit system in `yt` knows how to handle units like `Mpc` by looking up unit
symbols with and without prefixes and modify the conversion factor appropriately. Note also that in most cases the "additive offset" will be 0--it applies mainly to temperature scales like Fahrenheit and Celsius where the zero-point of the scale are not equivalent between the two scales, and are also not equivalent to "zero thermal energy", for which the relevant temperature units are Rankine and Kelvin, respectively.

We construct a `Unit` object by providing a string containing atomic unit symbols, combined with operations in Python syntax, and the registry those atomic unit symbols are defined in. We use
`sympy`'s string parsing features to create the unit expression from the user-provided string. Each `Unit`

> object also has its own unit `sympy` `Expression` object, a dimensionality `Expression` object, a `UnitRegistry` instance, a scalar conversion factor to SI units, additive offset value, and display representation. Among other methods, `Unit` defines the `mul`, `div`, `pow`, and `eq` operations with

Reviewer: are `Expr` and `Expression` the same class ?

Author: Changed it to `Expr`

look up the symbols. The unit system in `yt` knows how to handle units like `Mpc` by looking up unit
symbols with and without prefixes and modify the conversion factor appropriately. Note also that in most cases the "additive offset" will be 0--it applies mainly to temperature scales like Fahrenheit and Celsius where the zero-point of the scale are not equivalent between the two scales, and are also not equivalent to "zero thermal energy", for which the relevant temperature units are Rankine and Kelvin, respectively.

We construct a `Unit` object by providing a string containing atomic unit symbols, combined with operations in Python syntax, and the registry those atomic unit symbols are defined in. We use
`sympy`'s string parsing features to create the unit expression from the user-provided string. Each `Unit`
object also has its own unit `sympy` `Expr` object, a dimensionality `Expr` object, a `UnitRegistry` instance, a scalar conversion factor to SI units, additive offset value, and display representation. Among other methods, `Unit` defines the `mul`, `div`, `pow`, and `eq` operations with
other unit objects, making it easy to compose compound units algebraically.

> The `UnitRegistry` class provides the `add`, `remove`, and `modify` methods which allows users to
> add, remove, and modify atomic unit definitions present in `UnitRegistry` objects. In general, unit registries should only be adjusted inside of a code frontend, since otherwise quantities and arrays might be created with inconsistent unit metadata. Once a `Unit` object is created, it will not receive updates if the original unit registry is modified.

Reviewer: Here's one possible way to avoid a repetition here, but it's not essential
```suggestion
The `UnitRegistry` class provides methods which allows users to
`add`, `remove`, and `modify` atomic unit definitions present in `UnitRegistry` objects. In general, unit registries should only be adjusted inside of a code frontend, since otherwise quantities and arrays might be created with inconsistent unit metadata. Once a `Unit` object is created, it will not receive updates if the original unit registry is modified.
```

#### Code units {#sec:code-units}

> When a `Dataset` object is instantiated, it will itself instantiate and set up a `UnitRegistry`

Reviewer: (cultural context: repetition is the bane of French teachers. This is probably why I'm chasing them, but I can imagine it's not that big of a deal in English)
```suggestion
When a `Dataset` object is created, it will itself instantiate and set up a `UnitRegistry`
```

[1.92e-31 1.74e-31 2.54e-31 ... 
 2.59e-28 1.61e-28 1.22e-28] g/cm**3
```

Here, the first data object query is returned in code units, while the second is returned in CGS
units. This is because `("enzo", "Density")` is an on-disk field, while `("gas", "density")` is an
internal `yt` field, aliased to the former. 

In the `set_code_units` method of the `Dataset` base class, the predefined `code_mass`,
`code_length`, `code_time`, `code_velocity`, and `code_magnetic` symbols are adjusted to the
appropriate values, and `length_unit`, `time_unit`, `mass_unit`, `velocity_unit`, and
`magnetic_unit` attributes are attached to the `Dataset` instance. If there are frontend-specific
code units, they should also be defined in subclasses by extending this function.

When code units are set up for a `Dataset`, `unyt_quantity` instances corresponding to these code units are also set up and added as attributes to the `Dataset`. Examples are `length_unit`, `time_unit`, `mass_unit`, and `temperature_unit`:

```python
>>> import yt
>>> ds = yt.load("Enzo_64/DD0043/data0043")
>>> print(ds.length_unit)
128 Mpccm/h
>>> print(ds.length_unit.in_cgs())
5.555172850264421e+26 cm
```

> Optionally, `velocity_unit`, `density_unit`, `pressure_unit`, and `magnetic_unit` may be defined as well,if the units for these fields cannot be inferred from the mass, length, and time units.

Reviewer: ```suggestion
Optionally, `velocity_unit`, `density_unit`, `pressure_unit`, and `magnetic_unit` may be defined as well, if the units for these fields cannot be inferred from the mass, length, and time units.
```

#### Handling cosmological units {#sec:handling-cosmological-units}

A special case of astrophysical hydrodynamical simulations are cosmological simulations, which
attempt to simulate the evolution of structure on cosmological length and time scales. Such
simulations use "comoving coordinates", which is a notion of length that does not change with the
expansion of the universe, in contrast to normal physical distances which do. 

If we detect that we are loading a cosmological simulation performed in comoving coordinates, extra
comoving units are added to the dataset's unit registry. Comoving length unit symbols are still
named following the pattern `<length symbol>cm`, i.e. `Mpccm`. The $h$ symbol is treated as a base
unit, `h`, which defaults to unity. The `Dataset` updates the `h` symbol to the correct value when
loading a cosmological simulation.

The change of physical length is illustrated by the following example, of two datasets from the same
simulation but at different times/scale factors/redshifts:

```python
>>> import yt

> >>> ds1 = yt.load('Enzo_64/DD0002/data0002')

Reviewer: this is where bug https://github.com/yt-project/yt/issues/4622 was discovered, right ?

Author: Yes, but I removed the offending code from this paper draft--I don't feel it's crucial to highlight the feature and I also didn't want to hold up finishing the paper before it was fixed.

units, `in_mks` for SI units, and the more general `in_base` for conversion to any valid unit
system (see [Unit systems](#sec:unit-systems)). 

```python
>>> import unyt as u 
>>> import numpy as np
>>> l = np.array([10, 100, 100])*u.Mpc
>>> print(l.in_cgs())
[3.08567758e+25 3.08567758e+26 3.08567758e+26] cm
>>> print(l.in_mks())
[3.08567758e+23 3.08567758e+24 3.08567758e+24] m
>>> print(l.in_base("imperial"))
[1.01236141e+24 1.01236141e+25 1.01236141e+25] ft
>>> print(l.in_base("geometrized"))
```

To convert an array in-place instead of making a copy, use `convert_to_units`:

```python
>>> momentum_y = sp["gas","mass"]*sp["gas","velocity_y"]
>>> print(momentum_y)
[1.45031068e+47 1.40467746e+47 1.42059875e+47 ... 1.01206589e+47
 9.94539437e+46 1.06969711e+47] cm*g/s
>>> momentum_y.convert_to_units("Msun*kpc/Myr")
>>> print(momentum_y)
[745945.76032301 722474.98923595 730663.86660289 ... 520541.05697294
 511526.58818102 550182.82006007] Msun*kpc/Myr
```

Naturally, attempting to convert an array to units with a different dimension raises an error (in
most cases, see below for exceptions):

> ```python

Reviewer: I know this "language" is supported by GitHub markdown, not sure about the flavour that's used here
```suggestion
```python-traceback
```

>>> print(momentum_y.to("degC"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/array.py", line 947, in to
    return self.in_units(units, equivalence=equivalence, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/array.py", line 874, in in_units
    (conversion_factor, offset) = self.units.get_conversion_factor(
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/unit_object.py", line 694, in get_conversion_factor
    return _get_conversion_factor(self, other_units, dtype)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/unit_object.py", line 939, in _get_conversion_factor
    raise UnitConversionError(
unyt.exceptions.UnitConversionError: Cannot convert between 'Msun*kpc/Myr' (dim '(length)*(mass)/(time)') and '°C' (dim '(temperature)').

> ```

Reviewer: I'm not sure what's happening but the view I'm getting in my browser doesn't seem to understand that this is where the code-block ends, which makes syntax highlighting completely useless in what follows. Maybe those are not the exact correct symbols ?

Author: Can you say which browser this happens in? I don't see the same issue.

Reviewer: Safari, but it's gone now, so maybe it was just a caching error !

desired. This is the case for a number of frequently encountered, physically-motivated
transformations which involve physical constants, some examples of which are:

* Conversions between temperature and energy via $E = k_BT$
* Conversions between wavelength, frequency, and energy for light via $E = h\nu = hc/\lambda$
* Conversions between mass and energy via $E = mc^2$
* Conversions between density and number density via $n = \rho/({\mu}m_p)$

As described above, conversions between quantities in different units is generally not allowed, but
these special cases can be handled using unit equivalencies in `yt`. This is done using the
`to_equivalent` method, which takes the unit in a different dimension that you want to convert to,
and the equivalence you would like to use: 

```python
>>> import unyt as u
>>> kT = 5.0*u.keV
>>> # converting kT (energy) to T (temperature)
>>> print(kT.to_equivalent("K", "thermal"))
34813557.843240075 K
>>> # converting kT (energy) to Angstrom (wavelength)
>>> print(kT.to_equivalent("angstrom", "spectral"))
4.132806438406553 Å
>>> # print the proton's compton wavelength in picometers 
>>> print(u.mp.to_equivalent("pm", "compton"))
0.00132141 pm
```

> These equivalencies are straightforward--in the case of others, more information may be required.

Reviewer: ```suggestion
These equivalencies are straightforward--more information may be required in other cases.
```

For example, to convert temperature to sound speed via $c_s = \sqrt{\gamma{k_BT}/(\mu{m_p})}$, there
are two parameters which may be optionally changed, $\gamma$ and $\mu$, which are the ratio of

> specific heats and the mean molecular weight, respectively:

Reviewer: ```suggestion
specific heat and the mean molecular weight, respectively:
```

Author: In this context, it is in fact "heats" (at least in English) because it's a ratio of C_p to C_v and each is a specific "heat"

Reviewer: oh, I must have misread it

                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/unit_object.py", line 694, in get_conversion_factor
    return _get_conversion_factor(self, other_units, dtype)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/unit_object.py", line 939, in _get_conversion_factor
    raise UnitConversionError(
unyt.exceptions.UnitConversionError: Cannot convert between 'A/m**3' (dim '(current_mks)/(length)**3') and 'statA/cm**3' (dim 'sqrt((mass))/((length)**(3/2)*(time)**2)').
```

For these cases, it is recommended to convert atomic electromagnetic units separately first in
equations, if necessary.

### Physical Constants

`unyt` contains a large number of built-in physical constants and 
properties of astronomical objects as `unyt_quantity` instances, which
can be used in computations. The simplest way to use them is to import
them directly from the `unyt` namespace:

```python
>>> from unyt import clight, kboltz, h, Mearth, G
>>> for const in [clight, kboltz, h, Mearth, G]:

> ...     print(const)

Reviewer: ```suggestion
...    print(const)
```

299792458.0 m/s
1.3806488e-23 kg*m**2/(K*s**2)
6.62606957e-34 kg*m**2/s
6.045644495102106e+24 kg
6.67408e-11 m**3/(kg*s**2)
```

Here's an example of using constants in a computation:
```python
>>> import numpy as np
>>> from unyt import unyt_quantity
>>> from unyt import Msun, G
>>> d = unyt_quantity(1.0, "AU")
>>> T = np.sqrt(4.0*np.pi**2*d**3/(G*Msun)) # Kepler's third law
>>> print(T.to("yr"))
1.000033863000043 yr
```

Some physical constants are represented by Greek letters or other characters--these can be imported either using ASCII characters or

> their non-ASCII representations:

Reviewer: It might be good to also demonstrate this part, as the non-ASCII names may not be obvious to the reader (they aren't to me)

Author: Done

new array and quantity objects: via a constructor, and by multiplying scalar data by a unit
quantity.

#### Class constructor {#sec:class-constructor}

The primary internal interface for creating new arrays and quantities is through the class
constructor for `unyt_array`. The constructor takes three arguments. The first argument is the input
data, which can be an integer, float, list, or array. The second argument, `input_units`, is a unit
specification which must be a string or `Unit` instance. Last, users may optionally supply a
`UnitRegistry` instance, which will be attached to the array. If no `UnitRegistry` is supplied, a
default unit registry is used instead. Unit specification strings must be algebraic combinations of
unit symbol names, using standard Python mathematical syntax (i.e. `**` for the power function, not
`^`).

Here is a simple example of `unyt_array` creation:

```python
>>> from unyt import unyt_array, unyt_quantity 
>>> unyt_array([1, 2, 3], 'cm') 
unyt_array([1, 2, 3]) cm
>>> unyt_quantity(3, 'J') 
unyt_quantity(3, 'J') 
```

In addition to the class constructor, we have also defined two convenience functions, `quan`, and

> `arr`, for quantity and array creation that are attached to the `Dataset` base class. These were

Reviewer: implementation detail
```suggestion
`arr`, for quantity and array creation that are attached to the `Dataset` class. These were
```

added to syntactically simplify the creation of arrays with the `UnitRegistry` instance associated
with a dataset. These functions work exactly like the `unyt_array` and `unyt_quantity` constructors,
but pass the `UnitRegistry` instance attached to the dataset to the underlying constructor call. For
example:

```python
>>> import yt
>>> ds = yt.load("Enzo_64/DD0043/data0043")
>>> ds.arr([1, 2, 3], 'code_length').in_cgs() 
unyt_array([5.55e+26, 1.11e+27, 1.66e+27], 'cm')
```

> This example illustrates that the array is being created using `ds.unit_registry`, rather than the `default_unit_registry` in `yt`.

Reviewer: ```suggestion
The above example illustrates that the array is being created using `ds.unit_registry`, rather than the `default_unit_registry` in `yt`.
```
